### PR TITLE
Not allowing disabling modeling checkers in ClangSA

### PR DESCRIPTION
**Pull request by dkrupp**
When a Clang Static Analyzer checker is disabled in CodeChecker, clang is invoked with the `analyzer-disable-checker` flag. This allows the user disabling core modeling checkers such as `unix.DynamicMemoryModeling`. This causes malfunctioning of depending checkers.

:exclamation: :exclamation: :exclamation: **Breaking Change:** :exclamation: :exclamation: :exclamation: 
After this patch, `modeling` and `debug` checkers (listed with `clang -cc1 -analyzer-checker-help-developer`)
will not be listed and cannot be disabled through CodeChecker with the `--enable` and `--disable` flags.

They can be enabled/disabled through the Clang Static Analyzer specific `--saargs` flag only.

### CODE REVIEW

**Review by csordasmarton — CHANGES_REQUESTED** [review]
Github action test cases are failed. I think the problem is that this option is supported properly only by the latest clang. So we should handle this on a version dependent way.

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py [hunk 1]
```diff
@@ -227,16 +228,23 @@ def construct_analyzer_cmd(self, result_handler):
...
-                    analyzer_cmd.extend(['-Xclang',
-                                         '-analyzer-checker=' + checker_name])
+                    enabled_checkers.append(checker_name)
                 elif state == CheckerState.disabled:
-                    analyzer_cmd.extend(['-Xclang',
-                                         '-analyzer-disable-checker=' +
-                                         checker_name])
+                    disabled_checkers.append(checker_name)
 
+            if len(enabled_checkers) > 0:
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py:240]:
```suggestion
            if enabled_checkers:
```

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py [hunk 2]
```diff
@@ -227,16 +228,23 @@ def construct_analyzer_cmd(self, result_handler):
...
+                analyzer_cmd.extend(['-Xclang',
+                                     '-analyzer-checker=' +
+                                     ','.join(enabled_checkers)])
+            if len(disabled_checkers) > 0:
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py:244]:
```suggestion
            if disabled_checkers:
```

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py [hunk 3]
```diff
@@ -135,7 +135,7 @@ def test_ctu_failure_zip(self):
         # with a specified name is analyzed.
         output, result = self.__do_ctu_all(
             on_demand=False, extra_args=["--verbose", "debug",
-                                         "-e", "debug.ExprInspection"])
+                                         "--saargs", "./ctu_failure.saargs"])
```
Comment (csordasmarton) [inline analyzer/tests/functional/ctu_failure/test_ctu_failure.py:138]: Can you please update the documentation and highlight this information that debug checkers can be enabled only by using --saargs and not with the --enable option?

Also this is a backward compatibility change that if someone enabled debug checkers with the --enable option from this release these checkers will not be enabled. Please highlight this information too in the docs and also in the messag of this PR.

File: analyzer/codechecker_analyzer/analyzers/clangsa/clang_options.py [hunk 4]
```diff
@@ -15,7 +15,7 @@
 LOG = get_logger('analyzer')
 
 
-def get_analyzer_checkers_cmd(cfg_handler, alpha=True, debug=True):
+def get_analyzer_checkers_cmd(cfg_handler, alpha=True, debug=False):
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/analyzers/clangsa/clang_options.py:18]: With this change we will not list `debug` checkers in the output of `CodeChecker checkers` command which is I think is not good. The purpose of this command is to list all the checkers supported by the analyzer. Debug checkers are also checkers and also can be turned on (from this patch by using --saargs only, but it is possible). So from the user perspective it would be good to list these checkers in the output too.

**Review by csordasmarton — CHANGES_REQUESTED** [review]
There is a typo in the commit message: `chekcers` vs `checkers`. Please update the commit message.

File: analyzer/codechecker_analyzer/analyzers/clangsa/clang_options.py (hunk 4 above)
Comment (dkrupp) [reply analyzer/codechecker_analyzer/analyzers/clangsa/clang_options.py:18]: End-Users should not use the debug checkers. Debug checkers should only be used by clang static analyzer developers and it is typically not used through codechecker, but rather in the lit test cases and for debugging checker implementations. I think it is enough if we allow users to enable these checkers from teh saargs and describe this as a workaround for edge cases. I will describe this procedure in https://github.com/Ericsson/codechecker/blob/master/docs/analyzer/checker_and_analyzer_configuration.md

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py (hunk 3 above)
Comment (dkrupp) [reply analyzer/tests/functional/ctu_failure/test_ctu_failure.py:138]: Done in docs/analyzer/checker_and_analyzer_configuration.md

File: docs/analyzer/checker_and_analyzer_configuration.md [hunk 5]
```diff
@@ -12,57 +38,22 @@ configuration options can be configured like this:
...
-| cfg-temporary-dtors                   | false             | true/false                               |                                                                                                     |
-| faux-bodies                           | true              | true/false                               |                                                                                                     |
-| graph-trim-interval                   | 1000              |                                          |                                                                                                     |
+### Enabling developer checkers
+You cannot enable/disable developer checkers in CodeChecker using the `--enable` or `--disable` flags.
 
+These (debug and modeling) checkers should not be used normally. The are typically used by ClangSA developers debug the analysis or to write test cases. These checkers can be listed by `clang -cc1 -analyzer-checker-help-developer`.
```
Comment (csordasmarton) [inline docs/analyzer/checker_and_analyzer_configuration.md:52]:
```suggestion
These (debug and modeling) checkers should not be used normally. They are typically used by ClangSA developers debug the analysis or to write test cases. These checkers can be listed by `clang -cc1 -analyzer-checker-help-developer`.
```

File: docs/analyzer/checker_and_analyzer_configuration.md [hunk 6]
```diff
@@ -1,7 +1,33 @@
...
+Some checkers can be customized using checker specific configuration options.
+
+These can be listed using the `CodeChecker checkers --checker-config` command and can be set by `CodeChecker analyze --checker-config clangsa:<option-name>=<value>`.
+
+You can find the documentation of the configuration options at hte 
+[Clang Static Analyzer checkers](https://github.com/llvm-mirror/clang/tree/master/lib/StaticAnalyzer/Checkers) page.
+
+## Clang Static Analyzer Special Configuration Options
+In special cases, when the checker and analyzer configurability that is provided by CodeChecker is not enough, the Clang Static analyzer configuration can be extended through the `--saargs` analysis option.
+The content of the saargs file are forwarded as  arguments without modification to the Clang Static Analyzer:
```
Comment (csordasmarton) [inline docs/analyzer/checker_and_analyzer_configuration.md:30]:
```suggestion
The content of the saargs file are forwarded as arguments without modification to the Clang Static Analyzer:
```

File: docs/analyzer/checker_and_analyzer_configuration.md [hunk 7]
```diff
@@ -97,14 +88,24 @@ command with the `--z3-refutation` option.
...
+## Configuring the analyzer and checkers
+
+You can configure the clang-tidy analyzer and its checkers through CodeChecker with the `--analyzer-config` and the `--checker-config`
+flags of `CodeChecker analyze/check` commands as described in sections [Analyzer Configuration](#analyzer-configuration) and [Checker Configuration](#checker-configuration).
+
 
 ## Using Clang tidy configuration files
 
-__clang-tidy__ attempts to read configuration for each analyzed source file
+If you want to control the configuration of clang tidy from the `.clang-tidy` configuration files (instead of the CodeChecker command line)  you can use the `clang-tidy:take-config-from-directory=true` option. It will skip setting the checkers and checker configuration from CodeChecker (even if a profile was specified).
```
Comment (csordasmarton) [inline docs/analyzer/checker_and_analyzer_configuration.md:101]: In the above section you are using `clang-tidy` but here you are using `clang tidy` without dash. Please be consistent:
```suggestion
If you want to control the configuration of clang-tidy from the `.clang-tidy` configuration files (instead of the CodeChecker command line)  you can use the `clang-tidy:take-config-from-directory=true` option. It will skip setting the checkers and checker configuration from CodeChecker (even if a profile was specified).
```

File: docs/analyzer/checker_and_analyzer_configuration.md [hunk 8]
```diff
@@ -97,14 +88,24 @@ command with the `--z3-refutation` option.
...
+
+Then __clang-tidy__ will attempt to read configuration for each analyzed source file
 from a `.clang-tidy` file located in the closest parent directory of the
 analyzed source file.
 
+So by executing `CodeChecker analyze compile_commands.json -o ./reports --analyzer-config 'clang-tidy:take-config-from-directory=true'`,  CodeChecker will generate a clang-tidy command which will NOT contain the -checks option at all so your .clang-tidy file will take precedence.
```
Comment (csordasmarton) [inline docs/analyzer/checker_and_analyzer_configuration.md:107]:
```suggestion
So by executing `CodeChecker analyze compile_commands.json -o ./reports --analyzer-config 'clang-tidy:take-config-from-directory=true'`, CodeChecker will generate a clang-tidy command which will NOT contain the `-checks` option at all so your .clang-tidy file will take precedence.
```

File: docs/analyzer/checker_and_analyzer_configuration.md [hunk 9]
```diff
@@ -1,7 +1,33 @@
 # Configure Clang Static Analyzer and checkers
 
-Analyzer configuration can be done through the `--saargs` analysis option which
-forwards arguments without modification to the Clang Static Analyzer:
+## Analyzer Configuration <a name="analyzer-configuration"></a>
+
+The analysis can be configured using analyzer wide configuration parameters. These parameters may have effect on the whole analysis, affecting the result of all checkers.
+
+Listing the avaialable configuration options:
```
Comment (csordasmarton) [inline docs/analyzer/checker_and_analyzer_configuration.md:7]:
```suggestion
Listing the available configuration options:
```

File: docs/analyzer/checker_and_analyzer_configuration.md [hunk 10]
```diff
@@ -1,7 +1,33 @@
 # Configure Clang Static Analyzer and checkers
 
-Analyzer configuration can be done through the `--saargs` analysis option which
-forwards arguments without modification to the Clang Static Analyzer:
+## Analyzer Configuration <a name="analyzer-configuration"></a>
+
+The analysis can be configured using analyzer wide configuration parameters. These parameters may have effect on the whole analysis, affecting the result of all checkers.
```
Comment (csordasmarton) [inline docs/analyzer/checker_and_analyzer_configuration.md:5]: Please break long lines in the markdown files too. The preferred max-length of each line should be 80 characters.

Note: I created an issue to introduce a linter for markdown files but it's not ready yet.

**Review by csordasmarton — CHANGES_REQUESTED** [review]

**Comment by dkrupp** [comment]
Thanks for the review. All comments fixed.

**Review by csordasmarton — APPROVED** [review]
LGTM!
